pkg/signup/service/signup_service.go

```diff
@@ -336,28 +336,77 @@ func (s *ServiceImpl) reactivateUserSignup(ctx *gin.Context, existing *toolchain
 // GetSignup returns Signup resource which represents the corresponding K8s UserSignup
 // and MasterUserRecord resources in the host cluster.
 // Returns nil, nil if the UserSignup resource is not found or if it's deactivated.
-func (s *ServiceImpl) GetSignup(userID, username string) (*signup.Signup, error) {
-	return s.DoGetSignup(s.defaultProvider, userID, username)
+func (s *ServiceImpl) GetSignup(ctx *gin.Context, userID, username string) (*signup.Signup, error) {
+	return s.DoGetSignup(ctx, s.defaultProvider, userID, username)
 }
 
 // GetSignupFromInformer uses the same logic of the 'GetSignup' function, except it uses informers to get resources.
 // This function and the ResourceProvider abstraction can replace the original GetSignup function once it is determined to be stable.
-func (s *ServiceImpl) GetSignupFromInformer(userID, username string) (*signup.Signup, error) {
-	return s.DoGetSignup(s.Services().InformerService(), userID, username)
+func (s *ServiceImpl) GetSignupFromInformer(ctx *gin.Context, userID, username string) (*signup.Signup, error) {
+	return s.DoGetSignup(ctx, s.Services().InformerService(), userID, username)
 }
 
-func (s *ServiceImpl) DoGetSignup(provider ResourceProvider, userID, username string) (*signup.Signup, error) {
-	// Retrieve UserSignup resource from the host cluster, using the specified UserID and username
-	userSignup, err := s.DoGetUserSignupFromIdentifier(provider, userID, username)
-	// If an error was returned, then return here
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil, nil
+func (s *ServiceImpl) DoGetSignup(ctx *gin.Context, provider ResourceProvider, userID, username string) (*signup.Signup, error) {
+	var userSignup *toolchainv1alpha1.UserSignup
+	var err error
+
+	err = pollUpdateSignup(ctx, func() error {
+		// Retrieve UserSignup resource from the host cluster, using the specified UserID and username
+		var getError error
+		userSignup, getError = s.DoGetUserSignupFromIdentifier(provider, userID, username)
+		// If an error was returned, then return here
+		if getError != nil {
+			if apierrors.IsNotFound(getError) {
+				return nil
+			}
+			return getError
+		}
+
+		// Otherwise if the returned userSignup is nil, return here also
+		if userSignup == nil || ctx == nil {
```

Comment:
Check if the ctx == nil as the first thing in the `pollUpdateSignup()` function. If it's nil then no need to get the usersignup in line 356.

Reply:
We still need the UserSignup even after the update function.

Reply:
Ah, right.